## Exercises for today

You will be doing 3 exercises to become familar with some of the bioinformatic tools that are used by researchers today. These tools are currently being used actively by researchers who are studying microbial genomes. 

### 1. sra-tools

The first is sra-tools, which are a suite of tools that you can use to download data from NCBI sequence read archives (SRA) (see here: https://www.ncbi.nlm.nih.gov/sra). This database is essential to a lot of researchers. This is where they will deposit raw sequences that came out of sequencing instruments. You can search for research project focusing on a certain topic and download the data they have deposited to SRA. It is publicly accessible. 

For example, if you type "Yellowstone hot spring" in the search box, you will see a list of accessions with links to the data. An example is here:

https://www.ncbi.nlm.nih.gov/sra/SRX9071323

If you click on that link, you will see that the study is titled "Seasonal hydrologic and geologic forcing drive hot spring geochemistry and microbial biodiversity". And there is a link below that says "Run, # of spots, # of Bases" etc. To download this sequence data, you will need to click on the link that starts with "SRR", click on the "Data access" tab, then click on the link provided next to the "run". 

So it becomes very cumbersome and difficult to navigate these pages to access and download the data. This is where the sra-tools becomes very useful. You can download the raw sequence reads deposited to SRA by just typing the following command with the accession number next to the `[accn]`.

```
fasterq-dump --split-3 SRR12584454
```

The "SRR12584454" is the actual run accession number for that particular study. This is actually the number you need to enter when trying to download these raw sequences from NCBI SRA.

First, you might want to explore what `fasterq-dump` means by typing `fasterq-dump -h`. `-h` just means you are trying to bring up help menu to list the options available with this command.

In [1]:
%%bash
fasterq-dump -h


Usage:
  fasterq-dump <path> [options]

Options:
  -o|--outfile                     output-file 
  -O|--outdir                      output-dir 
  -b|--bufsize                     size of file-buffer dflt=1MB 
  -c|--curcache                    size of cursor-cache dflt=10MB 
  -m|--mem                         memory limit for sorting dflt=100MB 
  -t|--temp                        where to put temp. files dflt=curr dir 
  -e|--threads                     how many thread dflt=6 
  -p|--progress                    show progress 
  -x|--details                     print details 
  -s|--split-spot                  split spots into reads 
  -S|--split-files                 write reads into different files 
  -3|--split-3                     writes single reads in special file 
  --concatenate-reads              writes whole spots into one file 
  -Z|--stdout                      print output to stdout 
  -f|--force                       force to overwrite existing file(s) 
  -N|--rowid-as-n

As you can see, there are multiple options available with this tool and you can even specify things like memory limitation for this command to run and the number of threads you want to use with this tool. Before actually typing the command, I want you to make sure that you change your directory to "exercises" directory that I made you create last week. I am giving you an example of where I created the folders and organize them. 

In [4]:
%%bash
cd ~/workspace/
tree MicrobialGenomicsLab

MicrobialGenomicsLab
├── data
├── exercises
├── repositories
└── tools

4 directories, 0 files


Here, you will see that I created the 4 folders inside "MicrobialGenomicsLab" folder which resides under "workspace". Navigate into the exercises folder and run the `fasterq-dump` tool with the example command that I wrote earlier.

```bash
fasterq-dump --split-3 SRR12584454
```

The command will start downloading the raw sequence files from SRA to your "exercises" folder. It may take a few minutes and you will see the report printed to the screen once it's done. Usually it will not tell you what is going on but you can increase the verbosity of the screen output by typing like this:

```bash
fasterq-dump -v --split-3 SRR12584454
```

Once it's done, you can type `ls -la` to see what files the tool produced.

In [5]:
%%bash
cd ~/workspace/MicrobialGenomicsLab/exercises/
ls -la

total 165888
drwxr-xr-x 4 jsaw 982768932      128 Sep  9 09:05 .
drwxr-xr-x 6 jsaw 982768932      192 Sep  9 08:58 ..
-rw-r--r-- 1 jsaw 982768932 70168140 Sep  9 09:05 SRR12584454_1.fastq
-rw-r--r-- 1 jsaw 982768932 70168140 Sep  9 09:05 SRR12584454_2.fastq


Now, you should see 2 files being produced by the `fasterq-dump` tool. Both files end with a ".fastq" file extension. These are fastq-formatted files that can be observed/analyzed with tools like `fastqc` or `bbmap`. Now, try to see what the contents of these fastq files look like.

In [9]:
%%bash
cd ~/workspace/MicrobialGenomicsLab/exercises/
head -8 SRR12584454_1.fastq

@SRR12584454.1 1 length=301
GTGCCAGCCGCCGCGGTAATACCAGCCCCGCGAGTGGTCGGGACTCTTACTGGGCCTAAAGCGCCCGTAGCCGGCCCGACAAGTCACTCCTTAAAGACCCCGGCTCAACCGGGGGAATGGGGGTGATACTGTCGGGCTAGGGGGCGGAAGAGGCCAGCGGTACTCCCGGAGTAGGGGCGAAATCCTCAGATCCCGGGAGGACCACCAGTGGCGAAAGCGGCTGGCTAGAACGCGCCCGACGGTGGGGGGCGAAAGGCGGGGCAGAGAAAGGGATTAGAAAACCCTTGAGGTCAGATGGGAA
+SRR12584454.1 1 length=301
CCCCCGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGDGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG@FCFGGGGGGGGGGGGGGGGGGGAFGGGGGGGEGGCCGGGGGCCGGGGGGGGGECGCFFGGGGGGGEFGGGGGGGGC6:CGGCGEEGGGGGGGGGG=:8EEC6C:C:?C*2<:<A<959*:763**:,;)/:EG(1):*-(2/><C@)-((.<F483((2))0:9,;855*.-*((,)(--)+20+
@SRR12584454.2 2 length=301
GTGTCAGCAGCCGCGGTAATACGGAGGGTGCGAGCGTTACTCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTTGTTAAGTTGGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATTCCAAACTGACGAGCTAGAGTATGGTAGAGGGTGGTGGAATTTCCTGTGTAGCGGTGAAATGCGTAGATATAGGAAGGAACACCAGTGGCGAAGGCGACCACCTGGAATGATACTGACAGTGAGGTGCGAAAGCGGGGGGAGCAAACAGGATTAGATACCCCGGTAGTCCAGATCGGAA
+SRR125844

I typed `head -8 SRR12584454_1.fastq` to print the first 8 lines of the fastq file to the terminal screen. As you can see, the file stores a string of letters representing the bases of DNA and other weird characters. Try to see if you can see a pattern here. Each record for a single fragment of DNA is represented by 4 lines. 

The first line starts with a '@' followed by a string of alphabets and numbers. This should be unique to each sequence record. The actual DNA sequence is on the 2nd line. The third line starts with '+' and same identifier as the first line. The 4th line contains characters that represent sequence quality for each of the DNA bases shown on the 2nd line. This is very important information for tools like `fastqc` or `bbmap` that relies on this information to assess sequence quality.

To understand more about how fastq files are encoded, see here: https://en.wikipedia.org/wiki/FASTQ_format

You see 2 fastq files that end with "\_1" and "\_2" in their file names. The reason for that is Illumina sequences that produced these files is usually run to sequence two ends of a single DNA fragment and for each DNA fragment that was sequenced, you have two sequences that you know are physically linked and therefore we call them **"read pairs"**. Pairing information is very useful for genome assembly and mapping. We will come to that in later labs.

Now that you have 2 fastq files, what do we do with them? First, what would you do if you want to know how many sequences are in these files? You can type something like this:

```bash
grep -c "@SRR12584454" SRR12584454_1.fastq
grep -c "@SRR12584454" SRR12584454_2.fastq
```

And both should return the same numbers. I will show you the example below. These are relatively small files by Illumina sequencing standards (only about 67 Mbp) so it is ok to run this `grep` command. I would not recommend this with files larger than several gigabytes large. 

In [10]:
%%bash
cd ~/workspace/MicrobialGenomicsLab/exercises/
grep -c "@SRR12584454" SRR12584454_1.fastq
grep -c "@SRR12584454" SRR12584454_2.fastq


103842
103842


As you can see, I am searching for the pattern "@SRR12584454" in each fastq file because I know that this identifier is present in each of the header line of the sequences. `grep -c` just counts the occurrences of these identifiers instead of printing the matches to screen. I see that both fastq files contain 103842 sequences. This is very small number. Usually you will have millions of sequences per fastq file. So there is no way for you to manually inspect each and every one of these sequences for their quality. This is where the FastQC tool comes in.

### 2. FastQC

FastQC tool can be found here: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/ but also through `conda` and you should have installed it through `conda` by now.

It is a tool to analyze high-throughput sequencing data such as those produced by Illumina sequencing technology. It exists in both graphical and command line modes. After installation, if you just type `fastqc` without any parameters, it will bring up the graphical interface. For bioinformaticians, however, we like to work in command line mode as much as possible due to large number of files that we usually need to process in automated fashion. Today, we will use fastqc to inspect the sequence quality of these 2 files you just downloaded from SRA.

In Unix environment, using this command on multiple sequences becomes easier because you can use wild card characters. For example, if you want to run fastqc and create reports that can be viewed, you can just type like this:

```bash
fastqc *.fastq
```

This means I am telling the `fastqc` tool to process any files in a given directory that ends with ".fastq" extension. I will show this below.

In [12]:
%%bash
cd ~/workspace/MicrobialGenomicsLab/exercises/
fastqc *.fastq

Analysis complete for SRR12584454_1.fastq
Analysis complete for SRR12584454_2.fastq


Started analysis of SRR12584454_1.fastq
Approx 5% complete for SRR12584454_1.fastq
Approx 10% complete for SRR12584454_1.fastq
Approx 15% complete for SRR12584454_1.fastq
Approx 20% complete for SRR12584454_1.fastq
Approx 25% complete for SRR12584454_1.fastq
Approx 30% complete for SRR12584454_1.fastq
Approx 35% complete for SRR12584454_1.fastq
Approx 40% complete for SRR12584454_1.fastq
Approx 45% complete for SRR12584454_1.fastq
Approx 50% complete for SRR12584454_1.fastq
Approx 55% complete for SRR12584454_1.fastq
Approx 60% complete for SRR12584454_1.fastq
Approx 65% complete for SRR12584454_1.fastq
Approx 70% complete for SRR12584454_1.fastq
Approx 75% complete for SRR12584454_1.fastq
Approx 80% complete for SRR12584454_1.fastq
Approx 85% complete for SRR12584454_1.fastq
Approx 90% complete for SRR12584454_1.fastq
Approx 95% complete for SRR12584454_1.fastq
Started analysis of SRR12584454_2.fastq
Approx 5% complete for SRR12584454_2.fastq
Approx 10% complete for SRR12584454_2.fast

This is what you would see if you type the commands in your terminal. You can now inspect what files are being produced after the `fastqc` command was run.

In [14]:
%%bash
cd ~/workspace/MicrobialGenomicsLab/exercises/
ls -la

total 169600
drwxr-xr-x 8 jsaw 982768932      256 Sep  9 09:41 .
drwxr-xr-x 6 jsaw 982768932      192 Sep  9 08:58 ..
-rw-r--r-- 1 jsaw 982768932 70168140 Sep  9 09:05 SRR12584454_1.fastq
-rw-r--r-- 1 jsaw 982768932   974550 Sep  9 09:41 SRR12584454_1_fastqc.html
-rw-r--r-- 1 jsaw 982768932   881128 Sep  9 09:41 SRR12584454_1_fastqc.zip
-rw-r--r-- 1 jsaw 982768932 70168140 Sep  9 09:05 SRR12584454_2.fastq
-rw-r--r-- 1 jsaw 982768932   973629 Sep  9 09:41 SRR12584454_2_fastqc.html
-rw-r--r-- 1 jsaw 982768932   867986 Sep  9 09:41 SRR12584454_2_fastqc.zip


As you can see, the command produced an "html" and a "zip" file for each fastq. Now, open the html files using your web browser. You should see something similar to this:

![fastqc](images/fqc.png)

This is just a screenshot of the very beginning of the report. As you scroll down, you will see more information. Basically, the check marks under "Summary" will tell you whether the sequences are good or bad. If you see multiple red crosses, that means the sequences are of poor or bad quality and shouldn't be used right away until further processing is done. An example of a good and a bad sequencing run are shown here:

https://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html

https://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html

A few things to note about the histograms in the first plot. You can see that each sequence is 301 basepair long and the histogram is basically trying to depict average sequence quality within a given window of sequence. You will notice that average sequence quality starts to drop as you go towards the end of the sequence. This is due to the nature of Illumina sequencing technology. It uses fluorescent molecules to record unique bases (A, C, G, T), and the fluorescence signal fades towards the later cycles of sequencing. This makes it difficult to confidently assign correct letters of DNA towards the end and the instrument records lower read qualities near the end.

Another thing you want to watch out for is the presence of "adapter" sequences (near the bottom of the report). See an example of bad sequence report. If the adapter sequences (which are artificial DNA constructs to facilitate sequencing) are left in the sequences for whatever reason, `fastqc` will detect it. In this bad sequence example, you will notice that the adapter content histogram starts to increase near the end of the sequences. If you see something like this, you will need to remove the adapter sequences before the sequences can be used in genome assemblies or other analyses.

This is where the `bbmap` tool comes in.

A very useful website on how to read and interpret sequencing quality assessments: https://sequencing.qcfail.com/articles/?report=reader

### 3. BBTools

BBTools is a collection of tools written by scientists at the Joint Genome Institute (JGI) in Berkeley, CA. See here: https://jgi.doe.gov/data-and-tools/bbtools/

It contains a number of tools that can perform tasks such as interconversion of file formats, processing of raw sequencing files into usable ones, mapping of sequence reads to reference genomes, etc. The tool we will be using today is `bbduk`, which is meant for filtering and trimming of reads for adapter, contaminants, and quality using k-mers. Before you can actually use this tool, you need a reference file containing all known adapter sequences. `bbduk` can then look up for these sequences to know if it can find these contaminants in your sequences.

First, download this adapter file here:

https://www.dropbox.com/s/f5mydteoupt8ugb/adapters.fa?dl=0

I suggest you put this "adapter.fa" file somewhere safe where there is no likelihood of it being deleted by accident.

Now, you can use `bbduk` to perform quality trimming and contaminant removal. To see what options are available with `bbduk`, type:

`bbduk.sh -h`

In [15]:
%%bash
bbduk.sh -h


Written by Brian Bushnell
Last modified March 24, 2020

Description:  Compares reads to the kmers in a reference dataset, optionally 
allowing an edit distance. Splits the reads into two outputs - those that 
match the reference, and those that don't. Can also trim (remove) the matching 
parts of the reads rather than binning the reads.
Please read bbmap/docs/guides/BBDukGuide.txt for more information.

Usage:  bbduk.sh in=<input file> out=<output file> ref=<contaminant files>

Input may be stdin or a fasta or fastq file, compressed or uncompressed.
If you pipe via stdin/stdout, please include the file type; e.g. for gzipped 
fasta input, set in=stdin.fa.gz

Input parameters:
in=<file>           Main input. in=stdin.fq will pipe from stdin.
in2=<file>          Input for 2nd read of pairs in a different file.
ref=<file,file>     Comma-delimited list of reference files.
                    In addition to filenames, you may also use the keywords:
                    adapters, artifacts, 

As you can see, the tool was written by Brian Bushnell and there are a lot of options you can specify with this tool. It can be pretty overwhelming to read and understand each of the parameter available for you to use. So I am giving you an example command with parameters I use for routing processing of raw sequence files.

```bash
bbduk.sh ktrim=r ordered minlen=50 mink=11 tbo rcomp=f k=21 ow=t ftm=5 zl=4 \
        qtrim=rl trimq=20 \
        in1=SRR12584454_1.fastq \
        in2=SRR12584454_2.fastq \
        ref=adapters.fa \
        out1=SRR12584454_1.trimmed.fastq \
        out2=SRR12584454_2.trimmed.fastq
```

Here you will notice a few things that are important in this example. First, I am telling `bbduk` to only keep sequences longer than 50 bases and quality higher than 20. And I am specifying the adapter file in `ref=adapters.fa` flag. `in1` and `in2` are where you specify input fastq files and `out1` and `out2` are where you specify output files, which should be named differently so that the program doesn't overwrite the original files. I will show what it looks like when you type this command.

In [16]:
%%bash
cd ~/workspace/MicrobialGenomicsLab/exercises/
bbduk.sh ktrim=r ordered minlen=50 mink=11 tbo rcomp=f k=21 ow=t ftm=5 zl=4 \
        qtrim=rl trimq=20 \
        in1=SRR12584454_1.fastq \
        in2=SRR12584454_2.fastq \
        ref=adapters.fa \
        out1=SRR12584454_1.trimmed.fastq \
        out2=SRR12584454_2.trimmed.fastq

/Users/jsaw/miniconda3/opt/bbmap-38.86-0//calcmem.sh: line 75: [: -v: unary operator expected
Max memory cannot be determined.  Attempting to use 1400 MB.
If this fails, please add the -Xmx flag (e.g. -Xmx24g) to your command, 
or run this program qsubbed or from a qlogin session on Genepool, or set ulimit to an appropriate value.
java -ea -Xmx1400m -Xms1400m -cp /Users/jsaw/miniconda3/opt/bbmap-38.86-0/current/ jgi.BBDuk ktrim=r ordered minlen=50 mink=11 tbo rcomp=f k=21 ow=t ftm=5 zl=4 qtrim=rl trimq=20 in1=SRR12584454_1.fastq in2=SRR12584454_2.fastq ref=adapters.fa out1=SRR12584454_1.trimmed.fastq out2=SRR12584454_2.trimmed.fastq
Executing jgi.BBDuk [ktrim=r, ordered, minlen=50, mink=11, tbo, rcomp=f, k=21, ow=t, ftm=5, zl=4, qtrim=rl, trimq=20, in1=SRR12584454_1.fastq, in2=SRR12584454_2.fastq, ref=adapters.fa, out1=SRR12584454_1.trimmed.fastq, out2=SRR12584454_2.trimmed.fastq]
Version 38.86

Set ORDERED to true
maskMiddle was disabled because useShortKmers=true
0.038 seconds.
Initi

When the tool finishes this task (which is pretty fast), you will see that it found 207,684 reads (if you combine the 2 fastq files) and removed 12,800 reads that do not meet the criteria I have set. And it retained 93.84% of the reads. The whole process only took less than 7 seconds. Now, if you look into the folder, you will notice 2 new files being produced that end with "`.trimmed.fastq`".

In [18]:
%%bash
cd ~/workspace/MicrobialGenomicsLab/exercises/
ls -lah

total 238M
drwxr-xr-x 11 jsaw 982768932  352 Sep  9 10:29 .
drwxr-xr-x  6 jsaw 982768932  192 Sep  9 08:58 ..
-rw-r--r--  1 jsaw 982768932  67M Sep  9 09:05 SRR12584454_1.fastq
-rw-r--r--  1 jsaw 982768932  53M Sep  9 10:29 SRR12584454_1.trimmed.fastq
-rw-r--r--  1 jsaw 982768932 952K Sep  9 09:41 SRR12584454_1_fastqc.html
-rw-r--r--  1 jsaw 982768932 861K Sep  9 09:41 SRR12584454_1_fastqc.zip
-rw-r--r--  1 jsaw 982768932  67M Sep  9 09:05 SRR12584454_2.fastq
-rw-r--r--  1 jsaw 982768932  44M Sep  9 10:29 SRR12584454_2.trimmed.fastq
-rw-r--r--  1 jsaw 982768932 951K Sep  9 09:41 SRR12584454_2_fastqc.html
-rw-r--r--  1 jsaw 982768932 848K Sep  9 09:41 SRR12584454_2_fastqc.zip
-rwxr-x---  1 jsaw 982768932  14K Sep  9 10:28 adapters.fa


The next step after this is to check if sequence qualities have improved after you have run `bbduk` tool. To do that, you run `fastqc` tool again but on the trimmed fastq files. Type:

```bash
fastqc *.trimmed.fastq
```

And inspect the sequence qualities by opening the html files produced by `fastqc`. Noticed any differences? You can open the `fastqc` report of the original file in another tab to see how different things are. You should see something like this:

![trimmed](images/fqct.png)

As you might notice, the histogram denoting sequence quality have drastically improved (no more columns dipping into the red zone). Now, the sequences are good to be used in downstream processes.

### 4. Preparations for next week's lab

We will be doing genome and metagenome assemblies next week. For that, you will download a few publicly available datasets and also familiarize yourself with using `ssh` and `rsync` tools to remotely connect to and transfer files to other computers. You can look at this page for some examples of how to use `ssh`.

https://phoenixnap.com/kb/linux-ssh-commands



Another tool that will be crucial for you to use next week will be this tool known as `rsync`. It is a command line tool that can be used to transfer files between two remote computers or even for syncing directories inside the same computer. You can see some examples of how to use `rsync` here:

https://www.tecmint.com/rsync-local-remote-file-synchronization-commands/

and here:

https://phoenixnap.com/kb/rsync-command-linux-examples

Try to play around with a few examples shown in these pages to become familar with how `rsync` works.

#### Download data for next week

We will be using publicly available sequence data from NCBI SRA and see if you can replicate their results. The first one to download is here:

https://www.ncbi.nlm.nih.gov/sra/SRX9094324

These are Illumina MiSeq sequences they have used to reconstruct the genome of a *Salmonella enteria* strain. This is a cultured isolate and represents a single organism. Use `fasterq-dump` tool to download this data. I recommend you put it in `data` folder on your computer. These are fairly large files (each about 450 MB). In order to save space, you can compress them after they are downloaded. You can type like this:

```bash
gzip *.fastq
```

And this will compress your text files into much smaller zipped files with extension "*.gz". The file sizes are much smaller after being compressed (about 110-130 MB).

For metagenome assemblies, we will be downloading this file:

https://www.ncbi.nlm.nih.gov/sra/SRX4741377

Make sure you use `fasterq-dump` to download it. Again, I recommend you put it in `data` folder on your computer and also use `gzip` to compress the fastq files as these are very large files (each about 4.5 GB, a total of 9 GB). These compressed files can be used directly by both `fastqc` and `bbduk`.

Now you are ready for next week's lab.